### DISCUSSION TITLE
Overlay ports for host triplet only

### DISCUSSION THREAD

**Discussion by huangqinjin** [microsoft/vcpkg#31662] [General]
Installing `python3:arm64-android` will also install host package `python3:x64-linux-dynamic`, I would like to use an empty overlay port for host package `python3`.

Currently I have to separate the installation command into two:
```
vcpkg install python3:x64-linux-dynamic --overlay-ports=./vcpkg-system-packages
vcpkg install python3:arm64-android --host-triplet x64-linux-dynamic
```
The environment variable `VCPKG_OVERLAY_PORTS` cannot be used, otherwise `python3:arm64-android` is also empty. Is there a better way for this?

**Comment by Osyotr** [comment]
I think it's possible with overlays:
```cmake
if(VCPKG_CROSSCOMPILING)
    set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
    return()
endif()

# rest of the portfile
```
Also you probably want this: https://github.com/microsoft/vcpkg/pull/31621

**Reply by huangqinjin** [reply]
This requires copying the original portfile to `# rest of the portfile` section right? Then it is not a really overlay.

**Reply by dg0yt** [reply]
Overlays are always complete ports. You probably talk about an *empty* overlay which is a special use case.

**Reply by huangqinjin** [reply]
What I need is a port that is, empty for host and complete for other triplets.

**Reply by dg0yt** [reply]
You must do it youself in the same overlay port. Its checksum is part of vcpkg ABI hashing and artifact caching, regardless of triplet.

**Reply by huangqinjin** [reply]
That is why I am looking for a mechanism to dispatch overlays based on triplets. So that I don't need duplicate the original ports in overlays.

**Reply by huangqinjin** [reply]
Maybe adding triplet matching to registry is a good idea https://github.com/microsoft/vcpkg/discussions/31313.

**Comment by dg0yt** [comment]
Why do you want to do that?
You should use the same version of python for the host as for target. Doing a lot of rebuilds for #31621, I would say build time is not a real issue.

**Reply by dg0yt** [reply]
FTR cross builds simply don't work before #31621, and they will require the python executable from the host port with that PR.

**Reply by huangqinjin** [reply]
Not only python, any other packages like `protobuf`, `openblas` , that my system provides, I use empty overlayports. My system is a rolling release Linux distro, the system packages versions are almost always latest and vcpkg packages are also almost latest, so version is not a issue to me.

**Reply by dg0yt** [reply]
Okay... 
IMO this is very far away from the supported (tested) field. Even more for host tool bootstrapping requirements which vcpkg ports must resolve by relying on the port's installation in the host triplet. Such as #31621 adding a dependency on the python interpreter path in the host triplet. 
So you might save some build time, but you will add troubleshooting time, for errors the vcpkg community will not be able to reproduce.

**Reply by huangqinjin** [reply]
> relying on the port's installation in the host triplet

Yeah, in this case, I actually need do something in overlay port (e.g. symlink to system tools).

> but you will add troubleshooting time

I expect no, as such host dependencies are almost to acquire some native tools, breaking something is less often. Anyway, I have to pay it if I do that.

**Reply by dg0yt** [reply]
... "some native tools" may work, but other native tools in vcpkg might be specifically patched to make them work, or may not be installed at all in another distribution.

**Reply by huangqinjin** [reply]
> be specifically patched

Could you give an example?

> may not be installed at all in another distribution

For this case, building from source is a must.
